File: 2025_fase2/solutions/cifracesar_alex.py
import sys
import string

A = string.ascii_uppercase

def pos(letra): return (ord(letra.upper()) - 65) % 26

def decifrar(texto, chave):
    res = []
    for ch in texto:
        if ch.upper() in A:
            res.append(A[(pos(ch) - chave) % 26])
        else:
            res.append(ch)
    return "".join(res)

for linha in sys.stdin:
    linha = linha.strip("\n")
    if linha == "***":  # encerra
        break
    if len(linha) < 2 or not linha[-1].isalpha():
        continue

    c1, c2 = linha[-2], linha[-1]
    dif = (26 - ((pos(c1) - pos(c2)) % 26)) % 26
    base = "E" if dif == 9 else "R" if dif == 1 else None
    if base:
        chave = (pos(c2) - pos(base)) % 26
        print(decifrar(linha, chave))

File: 2025_fase2/solutions/cifracesar_alex.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#define MAX 1024

int pos(char letra) {
    return (toupper(letra) - 'A') % 26;
}

void decifrar(const char *texto, int chave) {
    for (int i = 0; texto[i]; i++) {
        char ch = texto[i];
        if (isalpha(ch)) {
            int novaPos = (pos(ch) - chave + 26) % 26;
            char novaLetra = 'A' + novaPos;
            putchar(isupper(ch) ? novaLetra : tolower(novaLetra));
        } else {
            putchar(ch);
        }
    }
    putchar('\n');
}

int main() {
    char linha[MAX];

    while (fgets(linha, sizeof(linha), stdin)) {
        linha[strcspn(linha, "\n")] = '\0'; // remove \n

        if (strcmp(linha, "***") == 0) break;
        int len = strlen(linha);
        if (len < 2 || !isalpha(linha[len-1])) continue;

        char c1 = linha[len-2], c2 = linha[len-1];
        int dif = (26 - ((pos(c1) - pos(c2) + 26) % 26)) % 26;

        char base = (dif == 9) ? 'E' : (dif == 1) ? 'R' : 0;
        if (base) {
            int chave = (pos(c2) - pos(base) + 26) % 26;
            decifrar(linha, chave);
        }
    }
    return 0;
}
